compiler/src/lnntoamm/Expr.ts

```diff
@@ -0,0 +1,316 @@
+      args.push(accessed);
+    }
+    args.push(...argAsts.map(a => {
+      let [generated, argExpr] = Expr.fromAssignablesAst(a, metadata);
+      stmts.push(...generated);
+      let arg: Ref;
+      if (argExpr instanceof Ref) {
+        arg = argExpr;
+      } else {
+        const dec = Dec.gen(argExpr, metadata);
+        stmts.push(dec);
+        arg = dec.ref();
+      }
+      return arg;
+    }));
+    let fns = metadata.scope.deepGet(fnName);
+    let closure = metadata.get(fnName);
+    if ((fns === null || !isFnArray(fns)) && closure === null) {
+      throw new Error(`no functions found for ${fnName}`);
+    }
+    if (fns === null || !isFnArray(fns)) {
+      fns = [] as Fn[];
+    }
+    let retPossibilities = [];
+    retPossibilities.push(...fns.map(fn => fn.retTy));
+    if (closure !== null) {
+      TODO('closures');
+    }
+    return [stmts, new Call(ast, fns, closure, args, Type.oneOf(retPossibilities))];
+  }
+
+  inline(amm: Output, kind: AssignKind, name: string, ty: Builtin) {
+    const argTys = this.args.map(arg => arg.ty.instance());
+    const selected = this.fns.reverse().find(fn => fn.acceptsTypes(argTys)) || null;
+    if (selected === null) {
+      throw new Error(`no function selected: ${this.ast.t.trim()}`);
+    }
+    selected.inline(amm, this.args, kind, name, ty);
+  }
+}
+
+class Const extends Expr {
+  val: string
+  private detectedTy: Builtin
+
+  get ty(): Type {
+    return this.detectedTy;
+  }
+
+  constructor(
+    ast: LPNode,
+    val: string,
+    detectedTy: Builtin,
+  ) {
+    super(ast);
+    this.val = val;
+    this.detectedTy = detectedTy;
+  }
+
+  static fromConstantsAst(
+    ast: LPNode,
+    _metadata: MetaData,
+  ): [Stmt[], Expr] {
+    let val = ast.t.trim();
+    let detectedTy = null;
+    if (ast.has('bool')) {
+      detectedTy = opcodes().get('bool');
+    } else if (ast.has('str')) {
+      detectedTy = opcodes().get('string');
+      // sanitize single-quoted strings
+      // don't need to for double-quoted strings, since the string output
+      // is double-quoted
+      if (val.startsWith("'")) {
+        let sanitized = val.substring(1, val.length - 1).replace(/'/g, "\\'");
+        val = `"${sanitized.replace(/"/g, '\\"')}"`;
+      }
+    } else if (ast.has('num')) {
+      if (val.indexOf('.') !== -1) {
+        detectedTy = Type.oneOf([
+          'float32',
+          'float64',
+        ].map(t => opcodes().get(t)));
+      } else {
+        detectedTy = Type.oneOf([
+          'float32',
+          'float64',
+          'int8',
+          'int16',
+          'int32',
+          'int64',
+        ].map(t => opcodes().get(t)));
+      }
+    } else {
+      throw new Error(`unrecognized constants node: ${ast}`);
+    }
+    return [[], new Const(ast, val, detectedTy)];
+  }
+
+  inline(amm: Output, kind: AssignKind, name: string, ty: Builtin) {
+    const globalName = amm.global('const', ty, this.val);
+    let copyOp = 'copy';
+    switch (ty.ammName) {
```

Comment:
Instead of a switch statement why not use something like: https://github.com/alantech/alan/pull/508/files#diff-6f21a0275e18e6835d1c7f6e90fba00333ab84951fe8205eeb29bed63861cf6eR23

```js
const getSuffix = (typename) => typename[0] + typename.replace(/[a-z]*/, '')
copyOp += getSuffix(ty.ammName) || ''
```